File: go/food-chain/food_chain.go
package foodchain

import "fmt"

var lines = []string{"It wriggled and jiggled and tickled inside her.\n", "How absurd to swallow a bird!\n", "Imagine that, to swallow a cat!\n", "What a hog, to swallow a dog!\n", "Just opened her throat and swallowed a goat!\n", "I don't know how she swallowed a cow!\n"}
var animals = []string{"fly", "spider", "bird", "cat", "dog", "goat", "cow"}

func Verse(v int) string {
	if v <= 0 || v > 8 {
		return ""
	} else if v == 8 {
		return "I know an old lady who swallowed a horse.\nShe's dead, of course!"
	} else {
		verse := fmt.Sprintf("I know an old lady who swallowed a %s.\n", animals[v-1])
		if v > 1 {
			verse += lines[v-2]
		}
		for i := v-2; i >= 0; i-- {
			if i == 1 {
				verse += "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
			} else {
				verse += fmt.Sprintf("She swallowed the %s to catch the %s.\n", animals[i+1], animals[i])
			}
		}
		return verse + "I don't know why she swallowed the fly. Perhaps she'll die."
	}
}

func Verses(start, end int) string {
	if start > end || start < 1 || end > 8 {
		return ""
	} else {
		verses := Verse(start)
		for i := start + 1; i <= end; i++ {
			verses += "\n\n" + Verse(i)
		}
		return verses
	}
}

func Song() string {
	return Verses(1, 8)
}

File: elixir/food-chain/lib/food_chain.ex
defmodule FoodChain do
  @creatures %{1 => "fly", 2 => "spider", 3 => "bird", 4 => "cat", 5 => "dog", 6 => "goat", 7 => "cow", 8 => "horse"}

  @lines %{
    1 => "",
    2 => "It wriggled and jiggled and tickled inside her.\n",
    3 => "How absurd to swallow a bird!\n",
    4 => "Imagine that, to swallow a cat!\n",
    5 => "What a hog, to swallow a dog!\n",
    6 => "Just opened her throat and swallowed a goat!\n",
    7 => "I don't know how she swallowed a cow!\n",
    8 => ""
  }

  @spider_line " that wriggled and jiggled and tickled inside her"

  @doc """
  Generate consecutive verses of the song 'I Know an Old Lady Who Swallowed a Fly'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) when start > stop, do: ""
  def recite(start, stop) do
    "#{verse(start)}#{if start != stop, do: "\n"}#{recite(start + 1, stop)}"
  end

  defp verse(number) do
    "#{beginning(number)}#{middle(number)}#{ending(number)}"
  end

  defp beginning(verse) do
    "I know an old lady who swallowed a #{@creatures[verse]}.\n#{@lines[verse]}"
  end

  defp middle(verse) when verse in [1, 8], do: ""
  defp middle(verse) do
    "She swallowed the #{@creatures[verse]} to catch the #{@creatures[verse-1]}#{if verse == 3, do: @spider_line}.\n#{middle(verse-1)}"
  end

  defp ending(8), do: "She's dead, of course!\n"
  defp ending(_), do: "I don't know why she swallowed the fly. Perhaps she'll die.\n"
end
